Support for not wrapping exceptions with TargetInvocationException.

AustinWise opened the pull request:
This add supports for the API described in dotnet/corefx#22866.

AustinWise commented:
There is at least one outstanding issue that might make it a bad idea to merge this pull request immediately:  dotnet/corert#4433 already added the new BindingFlags value to the shared System.Private.CoreLib code. I'm not sure what effect adding it this pull request will have. If it causes a problem, I will rebase once that change is merged.

Additionally, there are a couple of there things I'm not sure are correct about this pull requests:
* If it is ok to call `CallDescrWorkerWithHandler` from `RuntimeMethodHandle::InvokeMethod`. The only other method in CoreCLR that calls this is also in reflectioninvocation.cpp. Seeing only one other caller makes me a little nervous.
* If the units tests belong here or in corefx.
* If I put the units tests in the right place.
* What is the right `CLRTestPriority` to assign the unit tests.

AtsushiKan commented:
Thanks for picking this up, @AustinWise! I'll start looking at the change in more detail but some immediate answers to your questions:

> dotnet/corert#4437 already added the new BindingFlags value to the shared System.Private.CoreLib code. I'm not sure what effect adding it this pull request will have.

As long as you picked up the exact change (including whitespaces), there shouldn't be any problem. There'll be an automated PR that brings BindingFlags.cs over from corert in the next few days - if the change has already been made in CoreCLR by then, it'll just be a null commit for that PR.

>If the units tests belong here or in corefx.

They belong in corefx, but as it happens, I've already written tests for this feature that I was planning to PR this week since I wanted test coverage for corert and didn't know how long the coreclr side would remain up for grabs. (You can see a preview here: https://github.com/AtsushiKan/corefx/blob/sig-wip/src/System.Runtime/tests/System/Reflection/BindingFlagsDoNotPass.netcoreapp.cs). So you needn't worry about that.

>I deleted some instances of StackWalkMark that were never passed to to a native method and the associated DynamicSecurityAttribute annotations. Is there anything to worry about when removing these inline-supressing attributes?

I'm not sure if this is safe or not - cc @jkotas, @gkhanna79

AtsushiKan commented on src/mscorlib/src/System/RuntimeHandles.cs:
```diff
@@ -912,7 +912,7 @@ internal static Utf8String GetUtf8Name(RuntimeMethodHandleInternal method)
         [Diagnostics.DebuggerHidden]
         [MethodImplAttribute(MethodImplOptions.InternalCall)]
-        internal extern static object InvokeMethod(object target, object[] arguments, Signature sig, bool constructor);
+        internal extern static object InvokeMethod(object target, object[] arguments, Signature sig, bool constructor, bool doNotWrapExceptions);
```
I think it would be good to make the new argument a positive (`wrapExceptions`) to match the CreateInstance method. It's easier to follow without the negative.

AtsushiKan replied:
Other than the "doNotWrapExceptions" vs. "wrapExceptions" issue, this looks good to me.

jkotas commented:
> deleted some instances of StackWalkMark 

Looks good. Thanks for cleaning it up.

AustinWise commented:
I updated this PR in response to the feedback by changing the "wrapExceptions" parameter to be consistently positive. I also removed the units tests in deference to the future unit tests in CoreFX.

@AtsushiKan In looking at your tests, they cover most code paths, but there are a couple `Activator.CreateInstance` left unexercised. CoreCLR has separate code paths for constructors with and without parameters. Additionally, if a parameterless constructor does not throw when called, it is cached. Subsequent calls to constructor happen through a delegate and the `TargetInvoationException` wrapping happens in C#. I covered these cases in my tests here: https://github.com/AustinWise/coreclr/blob/692ab2810bb9bd4e81e9187230dd7d0439ef1c61/tests/src/reflection/DoNotWrapExceptions/DoNotWrapExceptions.cs#L127-L137

AtsushiKan commented:
@dotnet-bot test OSX10.12 x64 Checked Build and Test

AtsushiKan commented:
@AustinWise - Thanks for your contribution! I've merged your PR and will augment my tests as you've suggested.
